From 941072580ada6e5ba55cdc1331da8297ba20204e Mon Sep 17 00:00:00 2001
From: Matthias Balke <matthias.balke@googlemail.com>
Date: Mon, 17 Mar 2025 23:23:20 +0100
Subject: [PATCH] change debian.master release version

---
 debian.master/changelog | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/debian.master/changelog b/debian.master/changelog
index f829da84a..4ab805072 100644
--- a/debian.master/changelog
+++ b/debian.master/changelog
@@ -1,4 +1,4 @@
-linux (6.8.0-55.57) noble; urgency=medium
+linux (6.8.0-55radeonvbiosfromdisk.57) noble; urgency=medium
 
   * noble/linux: 6.8.0-55.57 -proposed tracker (LP: #2097981)
 
-- 
2.43.0

